### DISCUSSION TITLE
Does Rundler work with pre eip 1559 networks ?

### DISCUSSION THREAD

**Discussion by nabeel99** [alchemyplatform/rundler#625] [Q&A] [answered]
Hey Everyone,

I have been deep diving into the rundler source code, tracing the estimate userop gas rpc and the sendUserOp rpc, i was curious how the gas limits are actually calculated (pvg,verification and call gaslimit) and how appropriate fees values are calculated, i understand it is  mainly using the block base fee + priority fee(estimated from fee history mechanism). I plan onto use this bundler in a network which is pre eip-1559 meaning there is no base fee concept so i was a bit confused if the rundler does not support it outright? because as far as i understand its basic calls to get base fee would fail ? and there is no handling around it that would suggest rundler is meant to support legacy networks too ?

**Comment by dancoombs** [comment, marked as answer]
Hey! Rundler currently only supports networks with EIP-1559 gas fees. We are open for contribution if you'd like to take a shot at generalizing our fee handling to handle pre-1559 networks. Would require some significant changes to the `sim` crate mostly.

**Reply by nabeel99** [reply]
Hey @dancoombs  , i talked internally and I would be more than happy to make this contribution, could you guys provide a high level overview/ things that need to be changed, how should we proceed with this this, should i open a pr ?

**Reply by dancoombs** [reply]
Sure! The best way forward would be to fork the repo and open a PR with your changes.

The way that you indicate that a UO is pre-1559 in ERC-4337 is you set `maxFeePerGas == maxPriorityFeePerGas`

At a high level what you'll need to change:

* Add a config to `ChainSpec` to indicate that the network is pre-1559.
* In `FeeEstimator`:
  * Update the logic for pre-1559 network fee estimation, ensure that in the return value `maxFeePerGas == maxPriorityFeePerGas`
  * Functions to look at are `required_bundle_fees` and `required_op_fees`.
* In `PrecheckerImpl`: 
  * Check that `maxFeePerGas == maxPriorityFeePerGas` on the UO if the network is pre-1559.
  * Ensure that the fee prechecks work for pre-1559 networks
* In `BundleProposerImpl` and `BundleSenderImpl` and the various transaction sender implementations ensure that the transaction that is being submitted is pre-1559 compliant.

There are a bunch of places in the code where we refer to `base_fee`, I think for a pre-1559 network we can just set that to 0 and have the gasFee == priority fee (hence the `maxFeePerGas == maxPriorityFeePerGas`) above.
